### ISSUE TITLE
Missing rules for SpecialFunctions

### ISSUE THREAD

**Issue by gdalle** [chalk-lab/Mooncake.jl#412]
Hi there!
I saw this Discourse post benchmarking reverse-mode AD on a code with Bessel functions: https://discourse.julialang.org/t/speeding-up-zygote-autodiff-for-numerical-loop/123515/
I tried Mooncake on it but it was missing a rule for `bessely`. When I tried implementing it, I got a weird error message (can't reproduce it here since I'm not on my work computer), which I think may be linked to the partial implementation of `ChainRulesCore.rrule`:
https://github.com/JuliaMath/SpecialFunctions.jl/blob/ed6a36732712a71e99de397cca45a58432f22a0e/ext/SpecialFunctionsChainRulesCoreExt.jl#L97-L103
Wanna take a look? It would be fun to see how close we can get to Enzyme.

**Comment by willtebbutt** [comment]
Yeah, the problem is that Mooncake can't (currently) handle `NotImplemented` stuff from ChainRules. In order to be able to do it without risking dropping gradient information, we need to be able to prove that it's safe to do so. At the minute we don't have any mechanism to do this.

This is to say that the only way to support the bessel functions currently is to finish the implementation off.

**Comment by willtebbutt** [comment]
A good example is something like the method of `SpecialFunctions.gamma` found [here](https://github.com/JuliaMath/SpecialFunctions.jl/blob/ed6a36732712a71e99de397cca45a58432f22a0e/src/gamma_inc.jl#L1119) with associated `ChainRules.rrule` found [here](https://github.com/JuliaMath/SpecialFunctions.jl/blob/ed6a36732712a71e99de397cca45a58432f22a0e/ext/SpecialFunctionsChainRulesCoreExt.jl#L55).

If you inspect the rule, you will see that (for whatever reason), the authors of SpecialFunctions have not provided the gradient w.r.t. the first argument, and instead return whatever `ChainRulesCore.@not_implemented(INCOMPLETE_GAMMA_INFO)` produces, which IIRC is a `ChainRulesCore.NotImplemented`. This is useful in the situation that you actually don't need the gradient w.r.t. the first argument of this function because it is e.g. a hard-coded constant. Again, IIRC, `ChainRulesCore.NotImplemented` has the property that if you try to add it to anything, you'll get an error.

This works well in Zygote, because Zygote it defers accumulating gradients until the rule which consumes them is reached. This means that if no rule consumes the `ChainRulesCore.NotImplemented` (e.g. because it's the gradient w.r.t. a hard-coded constant), everything works correctly. Notice that there is no risk of dropping gradients incorrectly here -- you get an error if you try to do anything with the result.

In Mooncake we accumulate the results in an eager manner i.e. immediately after the reverse-pass of the rule. This makes the reverse-pass dramatically more simple in general, but has the unfortunate consequence in this case that we can't make use of the `NotImplemented` feature of ChainRules quite as straightforwardly. Consequently, we need some kind of mechanism to explicitly tell us that a `NotImplemented` gradient is safe to ignore. In principle we have this information, but I haven't started to think about how we would exploit it in a systematic way -- it feels to me like something that we should think about when we add activity analysis to Mooncake more generally.

**Comment by 3f6a** [comment]
Just came across this thread. Also `besselix` is missing, probably for the same reason (ChainRules does not give the derivative for the first argument).

If I understand the issue here, is that one would have to implement also the derivative with respect to the first argument so that Mooncake can pick it up?

> In Mooncake we accumulate the results in an eager manner i.e. immediately after the reverse-pass of the rule

This sounds inefficient, or you have a way of not doing computations that are not needed in the end? Sorry for the naive question, this is the first time I'm trying out Mooncake.

**Comment by AstitvaAggarwal** [comment]
@3f6a For the first part of the question, yes I'm actually working on this. I'm thinking for now I'll follow what [`DiffRules.jl`](https://github.com/JuliaDiff/DiffRules.jl/blob/8842177391b07dcd8234ac7612b9ca8ca72d28e0/src/rules.jl#L204) + `ForwardDiff.jl` nan safe mode end up allowing -> 0 as the returned gradient for the unhandled/intractable gradients in `SpecialFunctions.jl` for methods like `besselix` etc.

Regarding the second part, Mooncake is actually benchmarked to be more efficient than Zygote (Atleast in progams where there are no or less dead branches).

**Comment by 3f6a** [comment]
> Mooncake is actually benchmarked to be more efficient than Zygote (At least in progams where there are no or less dead branches)

What does this last part mean: `At least in progams where there are no or less dead branches` ?

**Comment by AstitvaAggarwal** [comment]
@3f6a A dead branch is the part of code that executes but has a zero derivative relative to your function's output. In an eager system like Mooncake, it's "dead weight" because the computer spends time calculating its gradient even though that gradient never changes the final outcome. Here activity analysis really helps to prune this "dead weight" while creating the final ruleset for the whole function.

So Mooncake works well for code which is usually more "efficiently" written.

**Comment by lassepe** [comment]
> I'm thinking for now I'll follow what DiffRules.jl + ForwardDiff.jl nan safe mode end up allowing -> 0 as the returned gradient for the unhandled/intractable gradients 

This seems like it would make for a very dangerous food-gun where Mooncake suddenly silently returns wrong gradients. I would prefer seeing`NaN` or `missing` for those elements.

**Comment by devmotion** [comment]
NaN-safe mode does not return incorrect gradients - on the contrary, non-NaN-safe mode may return wrong gradients. There are multiple examples of this incorrect behaviour in ForwardDiff and its use in eg Distributions and Turing. NaN-safe mode avoids that gradients are affected by parts of the computational graph that are not differentiated.

**Comment by yebai** [comment]
Fixed by https://github.com/chalk-lab/Mooncake.jl/pull/938
